sklearn/neighbors/base.py

```diff
@@ -773,11 +773,22 @@ def fit(self, X, y):
             self.outputs_2d_ = True
 
         check_classification_targets(y)
+
+        self._issparse = issparse(y)
+
         self.classes_ = []
-        self._y = np.empty(y.shape, dtype=np.int)
-        for k in range(self._y.shape[1]):
-            classes, self._y[:, k] = np.unique(y[:, k], return_inverse=True)
-            self.classes_.append(classes)
+        if self._issparse and self.outputs_2d_:
+            self._y = y
+            for k in range(self._y.shape[1]):
```

Comment:
I don't think `range` should be used here. Other than that, this is much better. Thanks.